accepted/future-releases/spread-collections/benchmarks/bin/profile_list_spread.dart

```diff
@@ -0,0 +1,135 @@
+import 'dart:collection';
+
+const trialMs = 100;
+const lengths = [0, 1, 2, 5, 10, 20, 50, 100, 1000];
+
+var csv = StringBuffer();
+
+class CustomList<T> extends ListBase<T> {
+  final List<T> _inner;
+
+  int get length => _inner.length;
+
+  set length(int value) => _inner.length = value;
+
+  CustomList(this._inner);
+
+  T operator [](int index) => _inner[index];
+
+  void operator []=(int index, T value) => _inner[index] = value;
+}
+
+void main() {
+  for (var length in lengths) {
+    var baseline = runBench("iterate", length, iterate);
+    runBench("List for", length, addList, baseline);
+    runBench("resize and set", length, resizeAndSet, baseline);
+    runBench("addAll()", length, addAll, baseline);
+    runBench("forEach()", length, forEach, baseline);
+    print("");
+  }
+
+//  print("");
+//  print(csv);
+}
+
+double runBench(
+    String name, int length, void Function(List<String>, List<String>) action,
+    [double baseline]) {
+  var from = <String>[];
+  for (var i = 0; i < length; i++) {
+    from.add(String.fromCharCode(i % 26 + 65));
+  }
+
+  var froms = [from, CustomList(from)];
+
+  var rate = benchBest(froms, action);
+
+  if (baseline == null) {
+    print("${length.toString().padLeft(4)} ${name.padRight(15)} "
+        "${rate.toStringAsFixed(2).padLeft(10)} spreads/ms "
+        "                 ${'-' * 20}");
+  } else {
+    var comparison = rate / baseline;
+    var bar = "=" * (comparison * 20).toInt();
+    if (comparison > 4.0) bar = "!!!";
+    print("${length.toString().padLeft(4)} ${name.padRight(15)} "
+        "${rate.toStringAsFixed(2).padLeft(10)} spreads/ms "
+        "${comparison.toStringAsFixed(2).padLeft(6)}x baseline $bar");
+  }
+
+  csv.writeln("$length,$name,$rate");
+  return rate;
+}
+
+/// Runs [bench] a number of times and returns the best (highest) result.
+double benchBest(
+    List<List<String>> froms, void Function(List<String>, List<String>) action) {
+  var best = 0.0;
+  for (var i = 0; i < 4; i++) {
+    var result = bench(froms, action);
+    if (result > best) best = result;
+  }
+
+  return best;
+}
+
+/// Spreads each list in [froms] into the middle of a list using [action].
+///
+/// Returns the number of times it was able to do this per millisecond, on
+/// average. Higher is better.
+double bench(List<List<String>> froms,
+    void Function(List<String>, List<String>) action) {
+  var elapsed = 0;
+  var count = 0;
+  var watch = Stopwatch()..start();
+  do {
+    for (var i = 0; i < froms.length; i++) {
+      var from = froms[i];
+      var to = <String>["a", "b"];
+
+      action(from, to);
+      to.add("b");
+      to.add("c");
+
+      count++;
+    }
+    elapsed = watch.elapsedMilliseconds;
```

Comment:
I'd recommend pulling this out of the loop.  I usually do something like:

```dart
var count = 1;
do {
  count *= 2;
  var watch = Stopwatch()..start();
  for (int i =0; i < count; i++) {
    ....// benchmark code
  }
  elapsed = watch.elapsedMilliseconds;
} while (elapsed < trialMs);
```

so that you're not calling the millisecond timer on every iteration.

Reply:
Yeah, it's a little dubious to me too, but it seems to be how the VM folks roll.

I'll land this as-is and then consider redoing how this works if I put more time into it.

Reply:
This is also how I do it, but I usually repeat the basic operation a number of times between checking elapsed:
```dart
do {
  for (var j = 0; j < 10000; j++) {  // number chosen so loop isn't too short or long.
     someTest();
     count += n;  // how many operations per test.
  }
  elapsed = watch.elapsedMilliseconds;
} while (elapsed < 2000);